### PR TITLE
Update allow result reuse

### PR DESCRIPTION
**Pull request by Horneth**
Unhardcode `allowResultReuse` and update its value when a cached job gets invalidated

- [x] Revert custom centaur branch

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala [hunk 1]
```diff
@@ -360,8 +365,21 @@ class EngineJobExecutionActor(replyTo: ActorRef,
...
 
   private def handleCacheInvalidatedResponse(response: CallCacheInvalidatedResponse, data: ResponsePendingData) = {
+    def updateMetadataForInvalidatedEntry(entry: CallCachingEntry) = {
+      import cromwell.services.metadata.MetadataService.implicits.MetadataAutoPutter
+      import cromwell.core.ExecutionIndex._
+      
+      val workflowId = WorkflowId.fromString(entry.workflowExecutionUuid)
+      // If the entry doesn't have an attempt, it means that this cache entry was added before this change
+      // and we don't know which attempt yielded this cache entry
+      // In that case make a best effort and update the first attempt
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala]: ToL: I would say a best effort would be to update the last attempt (the one most likely to have actually succeeded)

OTOH, I imagine that's probably horrible... can we update *all* attempts?

**Comment by cjllanwarne** [comment]
ToL: worth a centaur?
👍

File: database/migration/src/main/resources/changesets/add_attempt_in_call_caching_entry.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+    <changeSet id="add_attempt_in_call_caching_entry" author="tjeandet">
+        <addColumn tableName="CALL_CACHING_ENTRY">
+            <column name="JOB_ATTEMPT" type="INT" defaultValue="NULL">
+                <constraints nullable="true" />
+            </column>
+        </addColumn>
+    </changeSet>
+
+</databaseChangeLog>
```
Comment (mcovarr) [inline database/migration/src/main/resources/changesets/add_attempt_in_call_caching_entry.xml]: \n

File: database/sql/src/main/scala/cromwell/database/slick/CallCachingSlickDatabase.scala [hunk 3]
```diff
@@ -49,10 +49,12 @@ trait CallCachingSlickDatabase extends CallCachingSqlDatabase {
...
-                               (implicit ec: ExecutionContext): Future[Unit] = {
-    import cats.syntax.functor._
-    import cats.instances.future._
-    val action = dataAccess.allowResultReuseForCallCachingEntryId(callCachingEntryId).update(false)
-    runTransaction(action) void
+                               (implicit ec: ExecutionContext): Future[Option[CallCachingEntry]] = {
+    val action = for {
+      _ <- dataAccess.allowResultReuseForCallCachingEntryId(callCachingEntryId).update(false)
+      callCachingEntryOption <- dataAccess.callCachingEntriesForId(callCachingEntryId).result.headOption
+    } yield callCachingEntryOption
```
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/slick/CallCachingSlickDatabase.scala]: what's using this new return value?

**Comment by mcovarr** [comment]
:+1:

File: database/sql/src/main/scala/cromwell/database/slick/CallCachingSlickDatabase.scala (hunk 3 above)
Comment (Horneth) [inline database/sql/src/main/scala/cromwell/database/slick/CallCachingSlickDatabase.scala]: The EJEA so it knows which call to update in the metadata to change the value of `allowResultReuse`

File: backend/src/main/scala/cromwell/backend/standard/StandardCachingActorHelper.scala [hunk 4]
```diff
@@ -75,9 +74,7 @@ trait StandardCachingActorHelper extends JobCachingActorHelper {
 
     val fileMetadata = jobPaths.metadataPaths
 
-    val otherMetadata = Map(CallMetadataKeys.CallCaching + MetadataKey.KeySeparator + "allowResultReuse" -> true)
-
-    runtimeAttributesMetadata ++ fileMetadata ++ otherMetadata ++ nonStandardMetadata
+    runtimeAttributesMetadata ++ fileMetadata ++ nonStandardMetadata
```
Comment (geoffjentry) [inline backend/src/main/scala/cromwell/backend/standard/StandardCachingActorHelper.scala]: Is this going to get called a ton? presumably these maps are small but if it's going to get called a bunch in the aggregate it'd add up if `add` is linear for the default `Map` class (I'm not sure on that)

File: database/sql/src/main/scala/cromwell/database/sql/CallCachingSqlDatabase.scala [hunk 5]
```diff
@@ -15,5 +16,5 @@ trait CallCachingSqlDatabase {
                       (implicit ec: ExecutionContext): Future[Option[CallCachingJoin]]
 
   def invalidateCall(callCachingEntryId: Int)
-                    (implicit ec: ExecutionContext): Future[Unit]
+                    (implicit ec: ExecutionContext): Future[Option[CallCachingEntry]]
```
Comment (geoffjentry) [inline database/sql/src/main/scala/cromwell/database/sql/CallCachingSqlDatabase.scala]: @mcovarr At some point you were looking at monad transformers to "unsuck" our `Future[Option]`s (`https://www.youtube.com/watch?v=hGMndafDcc8`) and I know we dropped it due to something or other. What was the something or other?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheInvalidateActor.scala [hunk 6]
```diff
@@ -33,6 +34,6 @@ object CallCacheInvalidateActor {
 }
 
 sealed trait CallCacheInvalidatedResponse
-case object CallCacheInvalidatedSuccess extends CallCacheInvalidatedResponse
+case class CallCacheInvalidatedSuccess(maybeEntry: Option[CallCachingEntry]) extends CallCacheInvalidatedResponse
 case object CallCacheInvalidationUnnecessary extends CallCacheInvalidatedResponse
 case class CallCacheInvalidatedFailure(t: Throwable) extends CallCacheInvalidatedResponse
```
Comment (geoffjentry) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/CallCacheInvalidateActor.scala]: viva la \n

File: database/sql/src/main/scala/cromwell/database/sql/CallCachingSqlDatabase.scala (hunk 5 above)
Comment (mcovarr) [inline database/sql/src/main/scala/cromwell/database/sql/CallCachingSqlDatabase.scala]: They didn't fit the shape of our problem.  We have very different handling of `Success(None)` versus `Failure`.

File: backend/src/main/scala/cromwell/backend/standard/StandardCachingActorHelper.scala (hunk 4 above)
Comment (Horneth) [inline backend/src/main/scala/cromwell/backend/standard/StandardCachingActorHelper.scala]: It'll be called once per job I believe.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala (hunk 1 above)
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/EngineJobExecutionActor.scala]: Discussed in person

**Pull request merged**